File: packages/daemon/src/shell-env.ts
import { execFileSync } from "node:child_process";

const PROBE_TIMEOUT_MS = 5_000;

const ENV_NAMES = [
  "PATH",
  "SSH_AUTH_SOCK",
  "HOMEBREW_PREFIX",
  "HOMEBREW_CELLAR",
  "HOMEBREW_REPOSITORY",
  "XDG_CONFIG_HOME",
  "XDG_DATA_HOME",
] as const;

type EnvName = (typeof ENV_NAMES)[number];

function startMarker(name: string): string {
  return `__ATHING_ENV_${name}_START__`;
}

function endMarker(name: string): string {
  return `__ATHING_ENV_${name}_END__`;
}

function buildCaptureCommand(names: readonly string[]): string {
  return names
    .map((name) =>
      [
        `printf '%s\\n' '${startMarker(name)}'`,
        `printenv ${name} || true`,
        `printf '%s\\n' '${endMarker(name)}'`,
      ].join("; "),
    )
    .join("; ");
}

function extractValue(output: string, name: string): string | undefined {
  const start = output.indexOf(startMarker(name));
  if (start === -1) return undefined;
  const valueStart = start + startMarker(name).length;
  const end = output.indexOf(endMarker(name), valueStart);
  if (end === -1) return undefined;
  const value = output.slice(valueStart, end).replace(/^\r?\n/, "").replace(/\r?\n$/, "");
  return value.length > 0 ? value : undefined;
}

function shellCandidates(): string[] {
  const candidates: string[] = [];
  if (process.env["SHELL"]) candidates.push(process.env["SHELL"]);
  if (process.platform === "darwin") {
    candidates.push("/bin/zsh", "/bin/bash");
  } else {
    candidates.push("/bin/bash", "/bin/sh");
  }
  return [...new Set(candidates)];
}

function probeShell(shell: string): Partial<Record<EnvName, string>> {
  const output = execFileSync(shell, ["-ilc", buildCaptureCommand(ENV_NAMES)], {
    encoding: "utf8",
    timeout: PROBE_TIMEOUT_MS,
  });
  const result: Partial<Record<EnvName, string>> = {};
  for (const name of ENV_NAMES) {
    const value = extractValue(output, name);
    if (value !== undefined) result[name] = value;
  }
  return result;
}

export function installLoginShellEnv(): void {
  for (const shell of shellCandidates()) {
    try {
      const env = probeShell(shell);
      if (!env.PATH) continue;

      // PATH: overwrite with login shell value (superset of daemon's minimal PATH)
      process.env["PATH"] = env.PATH;

      // Other vars: only set if not already present
      for (const name of ENV_NAMES) {
        if (name === "PATH") continue;
        const value = env[name];
        if (value && !process.env[name]) {
          process.env[name] = value;
        }
      }
      return;
    } catch {
      continue;
    }
  }
}

File: packages/daemon-pty/src/shell_env.rs
//! Shell env: spawned commands resolve as in user terminal.

use std::process::Command;

const PROBE_TIMEOUT_SECS: u64 = 5;

const ENV_NAMES: &[&str] = &[
    "PATH",
    "SSH_AUTH_SOCK",
    "HOMEBREW_PREFIX",
    "HOMEBREW_CELLAR",
    "HOMEBREW_REPOSITORY",
    "XDG_CONFIG_HOME",
    "XDG_DATA_HOME",
];

fn start_marker(name: &str) -> String {
    format!("__ATHING_ENV_{name}_START__")
}
fn end_marker(name: &str) -> String {
    format!("__ATHING_ENV_{name}_END__")
}

fn build_capture_command() -> String {
    ENV_NAMES
        .iter()
        .map(|name| {
            format!(
                "printf '%s\\n' '{}'; printenv {} || true; printf '%s\\n' '{}'",
                start_marker(name),
                name,
                end_marker(name)
            )
        })
        .collect::<Vec<_>>()
        .join("; ")
}

fn extract_value(output: &str, name: &str) -> Option<String> {
    let start_m = start_marker(name);
    let end_m = end_marker(name);
    let start = output.find(&start_m)? + start_m.len();
    let end = output[start..].find(&end_m)? + start;
    let value = output[start..end]
        .trim_start_matches(['\r', '\n'])
        .trim_end_matches(['\r', '\n']);
    if value.is_empty() {
        None
    } else {
        Some(value.to_string())
    }
}

fn shell_candidates() -> Vec<String> {
    let mut candidates = Vec::new();
    if let Ok(s) = std::env::var("SHELL") {
        if !s.is_empty() {
            candidates.push(s);
        }
    }
    if cfg!(target_os = "macos") {
        candidates.push("/bin/zsh".into());
        candidates.push("/bin/bash".into());
    } else {
        candidates.push("/bin/bash".into());
        candidates.push("/bin/sh".into());
    }
    let mut seen = std::collections::HashSet::new();
    candidates
        .into_iter()
        .filter(|c| seen.insert(c.clone()))
        .collect()
}

fn probe_shell(shell: &str) -> Option<Vec<(String, String)>> {
    let _ = PROBE_TIMEOUT_SECS;
    let out = Command::new(shell)
        .args(["-ilc", &build_capture_command()])
        .output()
        .ok()?;
    let stdout = String::from_utf8_lossy(&out.stdout);
    let mut result = Vec::new();
    for name in ENV_NAMES {
        if let Some(value) = extract_value(&stdout, name) {
            result.push((name.to_string(), value));
        }
    }
    Some(result)
}

pub fn install_login_shell_env() {
    for shell in shell_candidates() {
        let Some(env) = probe_shell(&shell) else {
            continue;
        };
        let path = env
            .iter()
            .find(|(k, _)| k == "PATH")
            .map(|(_, v)| v.clone());
        let Some(path) = path else {
            continue;
        };
        std::env::set_var("PATH", &path);
        for (name, value) in &env {
            if name == "PATH" {
                continue;
            }
            if std::env::var_os(name).is_none() && !value.is_empty() {
                std::env::set_var(name, value);
            }
        }
        return;
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn extract_value_between_markers() {
        let out = format!(
            "{}\n/usr/bin:/bin\n{}\n",
            start_marker("PATH"),
            end_marker("PATH")
        );
        assert_eq!(
            extract_value(&out, "PATH").as_deref(),
            Some("/usr/bin:/bin")
        );
    }

    #[test]
    fn extract_missing_is_none() {
        assert_eq!(extract_value("nothing here", "PATH"), None);
    }
}
